### PR TITLE
feat(webpack-preprocessor): Allow to pass custom webpack output

### LINKED ISSUE #3491: ffmpeg: Error initializing output stream: Error while opening encoder for output stream: width not divisible by 2
### Current behavior:

Tests doesn't run without the GUI because of the error with video capture.

```
$ DEBUG=cypress:server:video node_modules/.bin/cypress run

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    3.1.4                                                                              │
  │ Browser:    Electron 59 (headless)                                                             │
  │ Specs:      1 found (ordering.spec.js)                                                         │
  │ Searched:   cypress/integration/ordering.spec.js                                               │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────

  Running: ordering.spec.js...                                                             (1 of 1)
  cypress:server:video capture started { command: 'ffmpeg -n 20 /home/yuri/.cache/Cypress/3.1.4/Cypress/resources/app/packages/server/node_modules/@ffmpeg-installer/linux-x64/ffmpeg -f image2pipe -use_wallclock_as_timestamps 1 -i pipe:0 -y -vcodec libx264 -preset ultrafast /srv/http/s1/cypress/videos/ordering.spec.js.mp4' } +0ms
  cypress:server:video capture stderr log { message: 'ffmpeg version N-45896-g19c3df0cd-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2018 the FFmpeg developers' } +3ms
  cypress:server:video capture stderr log { message: '  built with gcc 6.3.0 (Debian 6.3.0-18+deb9u1) 20170516' } +1ms
  cypress:server:video capture stderr log { message: '  configuration: --enable-gpl --enable-version3 --enable-static --disable-debug --disable-ffplay --disable-indev=sndio --disable-outdev=sndio --cc=gcc-6 --enable-fontconfig --enable-frei0r --enable-gnutls --enable-gray --enable-libaom --enable-libfribidi --enable-libass --enable-libvmaf --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librubberband --enable-libsoxr --enable-libspeex --enable-libvorbis --enable-libopus --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg' } +0ms
  cypress:server:video capture stderr log { message: '  libavutil      56. 17.100 / 56. 17.100' } +0ms
  cypress:server:video capture stderr log { message: '  libavcodec     58. 19.100 / 58. 19.100' } +1ms
  cypress:server:video capture stderr log { message: '  libavformat    58. 13.100 / 58. 13.100' } +0ms
  cypress:server:video capture stderr log { message: '  libavdevice    58.  4.100 / 58.  4.100' } +0ms
  cypress:server:video capture stderr log { message: '  libavfilter     7. 20.100 /  7. 20.100' } +0ms
  cypress:server:video capture stderr log { message: '  libswscale      5.  2.100 /  5.  2.100' } +0ms
  cypress:server:video capture stderr log { message: '  libswresample   3.  2.100 /  3.  2.100' } +0ms
  cypress:server:video capture stderr log { message: '  libpostproc    55.  2.100 / 55.  2.100' } +13ms
  cypress:server:video capture stderr log { message: '[mjpeg @ 0x6883000] EOI missing, emulating' } +370ms


  ordering
  cypress:server:video capture stderr log { message: 'Input #0, image2pipe, from \'pipe:0\':' } +7s
  cypress:server:video capture stderr log { message: '  Duration: N/A, start: 1550258710.960000, bitrate: N/A' } +0ms
  cypress:server:video capture stderr log { message: '    Stream #0:0: Video: mjpeg, yuvj420p(pc, bt470bg/unknown/unknown), 1023x575 [SAR 1:1 DAR 1023:575], 25 fps, 25 tbr, 25 tbn, 25 tbc' } +0ms
  cypress:server:video capture stderr log { message: 'Stream mapping:' } +1ms
  cypress:server:video capture codec data: { format: 'image2pipe', audio: '', video: 'mjpeg', duration: 'N/A', video_details: [ 'mjpeg', 'yuvj420p(pc', 'bt470bg/unknown/unknown)', '1023x575 [SAR 1:1 DAR 1023:575]', '25 fps', '25 tbr', '25 tbn', '25 tbc' ] } +0ms
  cypress:server:video capture stderr log { message: '  Stream #0:0 -> #0:0 (mjpeg (native) -> h264 (libx264))' } +3ms
  cypress:server:video capture stderr log { message: '[libx264 @ 0x6887a00] width not divisible by 2 (1023x575)' } +3ms
  cypress:server:video capture stderr log { message: 'Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height' } +0ms
  cypress:server:video capture stderr log { message: 'Conversion failed!' } +2ms
  cypress:server:video capture stderr log { message: '' } +1ms
  cypress:server:video capture errored: { error: 'ffmpeg exited with code 1: Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height\nConversion failed!\n', stdout: '', stderr: 'ffmpeg version N-45896-g19c3df0cd-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2018 the FFmpeg developers\n  built with gcc 6.3.0 (Debian 6.3.0-18+deb9u1) 20170516\n  configuration: --enable-gpl --enable-version3 --enable-static --disable-debug --disable-ffplay --disable-indev=sndio --disable-outdev=sndio --cc=gcc-6 --enable-fontconfig --enable-frei0r --enable-gnutls --enable-gray --enable-libaom --enable-libfribidi --enable-libass --enable-libvmaf --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librubberband --enable-libsoxr --enable-libspeex --enable-libvorbis --enable-libopus --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg\n  libavutil      56. 17.100 / 56. 17.100\n  libavcodec     58. 19.100 / 58. 19.100\n  libavformat    58. 13.100 / 58. 13.100\n  libavdevice    58.  4.100 / 58.  4.100\n  libavfilter     7. 20.100 /  7. 20.100\n  libswscale      5.  2.100 /  5.  2.100\n  libswresample   3.  2.100 /  3.  2.100\n  libpostproc    55.  2.100 / 55.  2.100\n[mjpeg @ 0x6883000] EOI missing, emulating\nInput #0, image2pipe, from \'pipe:0\':\n  Duration: N/A, start: 1550258710.960000, bitrate: N/A\n    Stream #0:0: Video: mjpeg, yuvj420p(pc, bt470bg/unknown/unknown), 1023x575 [SAR 1:1 DAR 1023:575], 25 fps, 25 tbr, 25 tbn, 25 tbc\nStream mapping:\n  Stream #0:0 -> #0:0 (mjpeg (native) -> h264 (libx264))\n[libx264 @ 0x6887a00] width not divisible by 2 (1023x575)\nError initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height\nConversion failed!\n' } +1ms
Warning: We failed to record the video.

This error will not alter the exit code.

Error: ffmpeg exited with code 1: Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height
Conversion failed!

    at ChildProcess.<anonymous> (/home/yuri/.cache/Cypress/3.1.4/Cypress/resources/app/packages/server/node_modules/fluent-ffmpeg/lib/processor.js:182:22)
    at emitTwo (events.js:125:13)
    at ChildProcess.emit (events.js:213:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)

{ Error: ffmpeg exited with code 1: Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height
Conversion failed!

    at ChildProcess.<anonymous> (/home/yuri/.cache/Cypress/3.1.4/Cypress/resources/app/packages/server/node_modules/fluent-ffmpeg/lib/processor.js:182:22)
    at emitTwo (events.js:125:13)
    at ChildProcess.emit (events.js:213:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
 recordingVideoFailed: true }
Error: ffmpeg exited with code 1: Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height
Conversion failed!

    at ChildProcess.<anonymous> (/home/yuri/.cache/Cypress/3.1.4/Cypress/resources/app/packages/server/node_modules/fluent-ffmpeg/lib/processor.js:182:22)
    at emitTwo (events.js:125:13)
    at ChildProcess.emit (events.js:213:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
```

```
$ xrandr
Screen 0: minimum 8 x 8, current 1024 x 576, maximum 32767 x 32767
eDP1 connected primary 1024x576+0+0 (normal left inverted right x axis y axis) 290mm x 170mm
   1920x1080     60.00 +  59.93    40.00  
   1680x1050     59.88  
   1400x1050     59.98  
   1600x900      60.00    59.95    59.82  
   1280x1024     60.02  
   1400x900      59.96    59.88  
   1280x960      60.00  
   1368x768      60.00    59.88    59.85  
   1280x800      59.81    59.91  
   1280x720      59.86    60.00    59.74  
   1024x768      60.00  
   1024x576      60.00*   59.90    59.82  
   960x540       60.00    59.63    59.82  
   800x600       60.32    56.25  
   864x486       60.00    59.92    59.57  
   640x480       59.94  
   720x405       59.51    60.00    58.99  
   640x360       59.84    59.32    60.00  
DP1 disconnected (normal left inverted right x axis y axis)
HDMI1 disconnected (normal left inverted right x axis y axis)
HDMI2 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
```

### Desired behavior:

Tests run.

And by the way, isn't it supposed to run all the tests even if it can't record the video?

### Steps to reproduce: (app code and test code)

Run the following test without GUI:

```js
describe('My First Test', function() {
  it('Visits the Kitchen Sink', function() {
    cy.wait(3000);   // or more to make it finish abnormally (prematurely)
    // low values still produce an error, but tests continue to run
  })
})
```

```
$ npx cypress run
```

If your screen resolution is greater than 1280x720. Otherwise:

```
$ xvfb-run -s '-screen 0 1280x720x24' npx cypress run
```

`electron` bug: [Screen dimensions are off by one pixel in offscreen mode](https://github.com/electron/electron/issues/21379).

### Workarounds

```
$ xvfb-run -s '-screen 0 1280x800x24' npx cypress run
```

Add `.outputOptions("-vf pad=ceil(iw/2)*2:ceil(ih/2)*2")` after the following [line][1]. Based on the following Stack Overflow [answer][2]. This fixes it for me.

Or probably `.videoFilter('pad=ceil(iw/2)*2:ceil(ih/2)*2')`.

[1]: https://github.com/cypress-io/cypress/blob/v3.1.5/packages/server/lib/video_capture.coffee#L75
[2]: https://stackoverflow.com/a/20848224/52499

### Versions

`cypress-3.1.4`, Arch Linux, Electron 59 (headless)

### LINKED ISSUE #14592: Mount Component that uses `React.lazy` throws webpack error
Test file:
```js
import { mount } from '@cypress/react';
...

describe('Test', function () {
  it('test test test', () => {
    mount(
      <Suspense fallback={<div>loading...</div>}>
        <MockedProvider mocks={[]}>
          <MemoryRouter>
            <TestComponent />
          </MemoryRouter>
        </MockedProvider>
      </Suspense>
    );
    expect(2 + 2).equal(4);
  });
```

This will work

```tsx
TestComponent.tsx

const TestComponent: FunctionComponent = () => {
	...
	return <div>...content...</div>
}
```

This won't work

```tsx
TestComponent.tsx

const TestComponent: FunctionComponent = () => {
	const Another = lazy(() => import('../../AnotherComponent'));
	...
	return <div>...content...</div>
}
```

even if this imported component is not in use.

Btw, I have `Suspense`

```js
const webpackOptions = {
  resolve: {
    extensions: ['.js', '.ts', '.tsx'],
    modules: [resolve(__dirname, 'src'), 'node_modules']
  },
  mode: 'development',
  devtool: false,
  output: {
    filename: 'bundle.js',
    path: resolve(__dirname, 'dist'),
    publicPath: '/'
  },
  module: {
    rules: [
      {
        test: /\.(ts|tsx)$/,
        use: [
          {
            loader: 'babel-loader',
            options: {
              rootMode: 'upward'
            }
          }
        ],
        exclude: /node_modules/
      },
      {
        test: /\.css$/,
        use: ['style-loader', 'css-loader', 'postcss-loader']
      }
    ]
  }
};
```

Thanks!

### PR DESCRIPTION
**Pull request by vtereshyn**
- Closes #14592

### User facing changelog
The user who uses `@cypress/webpack-preprocessor` has the ability to set his/her own Webpack `publicPath`

### Additional details
Please, see https://github.com/cypress-io/cypress/issues/14592 for additional explanation.
### How has the user experience changed?

Before: 
image

After:
image

### PR Tasks
- [+] Have tests been added/updated?

### CODE REVIEW

**Comment by vtereshyn** [comment]
Not sure who I have to tag in here, hope that @bahmutov will be the right person to address this.

**Review by dmtrKovalenko — CHANGES_REQUESTED** [review]
I am not sure it will work with lazy-loaded modules (cypress has some limitations here – only allows 1 javascript file). There is a test https://github.com/cypress-io/cypress/blob/develop/npm/react/cypress/component/advanced/lazy-loaded-suspense/lazy-loaded-suspense.spec.tsx that can check the desired behavior. 

Can you please uncomment it and make sure that CI passed? Thanks

**Comment by vtereshyn** [comment]
@dmtrKovalenko I am pretty sure it will work with Lazy loaded components as I showed in my screenshots.
 
Here is my usage: I have a component that wrapped into HOC which lazy loads another component (popup) and I got a Webpack error, not React or Cypress.

Right now as a fix I'm editing `node_modules/@cypress/` folder to proceed with my work. 

Also, sorry, I didn't get what particularly should I uncomment? I see that the code you mentioned is uncommented.

**Comment by dmtrKovalenko** [comment]
Cool if so, sorry just remove the `.skip` on test and the coment

**Comment by vtereshyn** [comment]
@dmtrKovalenko ok, will do. But my fix is not about that at all. My fix is particularly about webpack-preprocessor and webpack error, not about Cypress and React.Lazy. I didn't change any code in Cypress core

**Comment by dmtrKovalenko** [comment]
I understand this, but PR title says that React.lazy breaks the webpack-preprocessor code, right?

File: npm/webpack-preprocessor/test/unit/index.spec.js [hunk 1]
```diff
@@ -151,6 +151,20 @@ describe('webpack preprocessor', function () {
         })
       })
 
+      it('adds output options when user output options are non-default', function () {
```
Comment (dmtrKovalenko) [inline npm/webpack-preprocessor/test/unit/index.spec.js:154]: This test doesn't really check that when we have dynamic import somewhere that webpack doesn't crash. It only checks that specific options passed to webpack config. 
What do you think about adding more complex test that shows the desired behavior if @cypress/react tests won't work?

**Comment by vtereshyn** [comment]
Probably this is incorrect to name PR as I named it here. I named it `React lazy loads breaks webpack config` just because I faced this error during Lazy loading in one of my components. This PR should be named like `Allow Cypress user to pass own Webpack output configuration`. Basically, this fix is about dynamic loading as well, just because Webpack doesn't understand the path inside the code.
Ex:
`import(`../../../${YourComponentName}``

**Comment by dmtrKovalenko** [comment]
Cool, thx its more clear now. Then it should be not `fix:` but `feat:` as well

**Comment by vtereshyn** [comment]
> Cool, thx its more clear now. Then it should be not `fix:` but `feat:` as well

Agreed, do you want me to create new PR with using new branch?

**Comment by dmtrKovalenko** [comment]
I think just rename the title

**Comment by vtereshyn** [comment]
@dmtrKovalenko sounds good, thanks

**Comment by vtereshyn** [comment]
@dmtrKovalenko ok, so do you want me to do something else? Write Cypress tests?

**Comment by dmtrKovalenko** [comment]
It may be nice to add e2e tests in webpack-preprocessor that shows how it works, but maybe that is not needed. :) 
Probably let's wait for CI to pass

**Comment by vtereshyn** [comment]
Ok, thanks!

**Comment by vtereshyn** [comment]
Hey @dmtrKovalenko 
I'm sorry, but I can't get what's going on CI because I have executed `test` script locally and everything was good. Not sure what should I do next and how can I reproduce CI issue but locally. Thanks!

**Comment by vtereshyn** [comment]
image

**Comment by vtereshyn** [comment]
image

I am not sure whats wrong on CI then

**Comment by vtereshyn** [comment]
@dmtrKovalenko any thoughts?

**Comment by vtereshyn** [comment]
well, I am not getting any responses, so I'm gonna close this PR. @dmtrKovalenko @jennifer-shehane

**Comment by dmtrKovalenko** [comment]
Do not close it, I'll take a look once get some free time

**Comment by lmiller1990** [comment]
@vtereshyn when you run the tests locally you may need to build `npm/webpack-preprocessor` manually - either run `yarn watch` in `npm/webpack-preprocessor` or just with `yarn build`. That might explain why things work locally (you ran the tests without your change) but CI builds everything prior to running the tests.

**Comment by vtereshyn** [comment]
@lmiller1990 thanks for your suggestion. I did `yarn build` before running tests. And looks like I got all test passed locally again.

**Comment by lmiller1990** [comment]
Will take a look!

**Comment by lmiller1990** [comment]
I ran the tests locally and reproduced the error once. But then I ran them again and it all seemed fine. I'll try re-running CI and see what happens.

**Comment by vtereshyn** [comment]
@lmiller1990 thank you

**Pull request closed without merging**